metrics/metrics_collector.cjs

```diff
@@ -0,0 +1,243 @@
+        "node.gc.old.collections": 0,
+        "node.gc.old.pause.ns": 0,
+        "node.gc.young.collections": 0,
+        "node.gc.young.pause.ns": 0,
+    }
+}
+
+/**
+ * Increments the memory counters based on the values in the performance entry if the entry has NodeGCPerformanceDetail information.
+ * @param {MemoryCounters} memoryCounters
+ * @param {PerformanceEntry} performanceEntry
+ */
+function updateMemoryCounters(memoryCounters, performanceEntry) {
+    // only record entries with that contain a 'detail' object with 'kind' property
+    // since these are only available for NodeGCPerformanceDetail entries
+    if ('detail' in performanceEntry && 'kind' in performanceEntry.detail) {
+        const nsDuration = millisecondsToNanoseconds(performanceEntry.duration)
+        memoryCounters['node.gc.collections'] += 1
+        memoryCounters['node.gc.pause.ns'] += nsDuration
+        if (performanceEntry.detail.kind === constants.NODE_PERFORMANCE_GC_MINOR) {
+            memoryCounters['node.gc.young.collections'] += 1
+            memoryCounters['node.gc.young.pause.ns'] += nsDuration
+        } else {
+            memoryCounters['node.gc.old.collections'] += 1
+            memoryCounters['node.gc.old.pause.ns'] += nsDuration
+        }
+    }
+}
+
+/**
+ * Collects details about the event loop metrics using perf_hooks functionality.
+ * @param {EventLoopUtilization} eventLoopUtilization
+ * @param {IntervalHistogram} eventLoopHistogram
+ * @returns {EventLoopGauges}
+ */
+function captureEventLoopGauges(eventLoopUtilization, eventLoopHistogram) {
+    return  {
+        'node.eventloop.usage.percent': eventLoopUtilization.utilization,
+        'node.eventloop.delay.ms.median': nanosecondsToMilliseconds(eventLoopHistogram.percentile(50)),
+        'node.eventloop.delay.ms.p95': nanosecondsToMilliseconds(eventLoopHistogram.percentile(95)),
+        'node.eventloop.delay.ms.p99': nanosecondsToMilliseconds(eventLoopHistogram.percentile(99)),
+        'node.eventloop.delay.ms.max': nanosecondsToMilliseconds(eventLoopHistogram.max),
+    }
+}
+
+/**
+ * Converts the given value in milliseconds into nanoseconds
+ * @param {number} ms A millisecond value
+ * @return {number}
+ */
+function millisecondsToNanoseconds(ms) {
+    return ms * 1_000_000
+}
+
+/**
+ * Converts the given value in nanoseconds into seconds
+ * @param {number} ns A nanosecond value
+ * @return {number}
+ */
+function nanosecondsToMilliseconds(ns) {
+    return ns / 1_000_000
+}
+
+/**
+ * Sends the collected metrics to the given endpoint using a POST request.
+ * @param {URL} url
+ * @param {MetricsPayload} payload
+ * @returns {Promise<void>}
+ */
+function sendMetrics(url, payload) {
+    log(`Sending metrics to ${url.toString()}`)
+    return fetch(url, {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify(payload)
+    }).then(res => {
+        if (res.ok) {
+            log('Metrics sent successfully')
```

Comment:
Seems like this might be a bit noisy in customer logs? Maybe this should only log in a verbose/debug scenario?

Reply:
Ah, `log` sends to a debug logger.